src/cpp/fastdds/subscriber/DataReaderImpl.cpp

```diff
@@ -218,12 +225,412 @@ DataReaderImpl::~DataReaderImpl()
+}
+
+ReturnCode_t DataReaderImpl::take(
+        LoanableCollection& data_values,
+        SampleInfoSeq& sample_infos,
+        int32_t max_samples,
+        SampleStateMask sample_states,
+        ViewStateMask view_states,
+        InstanceStateMask instance_states)
+{
+    if (reader_ == nullptr)
+    {
+        return ReturnCode_t::RETCODE_NOT_ENABLED;
+    }
+
+    ReturnCode_t code = check_collection_preconditions_and_calc_max_samples(data_values, sample_infos, max_samples);
+    if (!code)
+    {
+        return code;
+    }
+
+    std::lock_guard<RecursiveTimedMutex> lock(reader_->getMutex());
+
+    code = prepare_loan(data_values, sample_infos, max_samples);
+    if (!code)
+    {
+        return code;
+    }
+
+    auto instance = history_.lookup_instance(HANDLE_NIL, false).second;
+    detail::StateFilter states{ sample_states, view_states, instance_states };
+    detail::ReadTakeCommand cmd(*this, data_values, sample_infos, max_samples, states, instance);
+    while (!cmd.is_finished())
+    {
+        cmd.add_instance(true);
+    }
+    return cmd.return_value();
+}
+
+ReturnCode_t DataReaderImpl::take_instance(
+        LoanableCollection& data_values,
+        SampleInfoSeq& sample_infos,
+        int32_t max_samples,
+        const InstanceHandle_t& a_handle,
+        SampleStateMask sample_states,
+        ViewStateMask view_states,
+        InstanceStateMask instance_states)
+{
+    if (reader_ == nullptr)
+    {
+        return ReturnCode_t::RETCODE_NOT_ENABLED;
+    }
+
+    ReturnCode_t code = check_collection_preconditions_and_calc_max_samples(data_values, sample_infos, max_samples);
+    if (!code)
+    {
+        return code;
+    }
+
+    std::lock_guard<RecursiveTimedMutex> lock(reader_->getMutex());
+
+    // Check if the instance exists
+    auto it = history_.lookup_instance(a_handle, true);
+    if (!it.first)
+    {
+        return ReturnCode_t::RETCODE_BAD_PARAMETER;
+    }
+
+    code = prepare_loan(data_values, sample_infos, max_samples);
+    if (!code)
+    {
+        return code;
+    }
+
+    detail::StateFilter states{ sample_states, view_states, instance_states };
+    detail::ReadTakeCommand cmd(*this, data_values, sample_infos, max_samples, states, it.second,
+            true);
+    while (!cmd.is_finished())
+    {
+        cmd.add_instance(true);
+    }
+    return cmd.return_value();
+}
+
+ReturnCode_t DataReaderImpl::take_next_instance(
+        LoanableCollection& data_values,
+        SampleInfoSeq& sample_infos,
+        int32_t max_samples,
+        const InstanceHandle_t& previous_handle,
+        SampleStateMask sample_states,
+        ViewStateMask view_states,
+        InstanceStateMask instance_states)
+{
+    if (reader_ == nullptr)
+    {
+        return ReturnCode_t::RETCODE_NOT_ENABLED;
+    }
+
+    ReturnCode_t code = check_collection_preconditions_and_calc_max_samples(data_values, sample_infos, max_samples);
+    if (!code)
+    {
+        return code;
```

Comment:
Missing negative test

Reply:
This should be checked by `collection_preconditions` test.